File: Palinderome Linked List/src/com/Releationship/LinkedList.java
package com.Releationship;

public class LinkedList {
    public Node reverse(Node head) {
        Node current = head;
        Node prev = null;

        while(current != null) {
            Node next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        head = prev;
        return head;
    }

    public boolean isPalindrome(Node head) {
        int length = 0;
        Node temp = head;

        while(temp != null) {
            temp = temp.next;
            length++;
        }
        temp = head;

        if(length % 2 == 0) {
            for (int i = 1; i < length/2; i++) {
                temp = temp.next;
            }
            Node second = reverse(temp);
            temp.next = null;
            Node first = head;
            while (first != null && second != null) {
                int val1 = first.data;
                int val2 = second.data;
                if (first.data != second.data) {
                    return false;
                }
                first = first.next;
                second = second.next;
            }
            return true;
        }
        else {
            for (int i = 1; i < length/2; i++) {
                temp = temp.next;
            }
            Node second = reverse(temp.next.next);
            temp.next = null;
            Node first = head;
            while (first != null && second != null) {
                if (first.data != second.data) {
                    return false;
                }
                first = first.next;
                second = second.next;
            }
            return true;
        }
    }
}

File: Check If Linked List is Palindrome/LinkedList.cpp
//
// Created by MRIGANKA on 3/14/2022.
//

#include <stack>
#include "LinkedList.h"

bool LinkedList::is_palindrome(Node *head) {
    std::stack<int> s;
    Node *temp = head;

    while (temp != nullptr) {
        s.push(temp->data);
        temp = temp->next;
    }

    temp = head;
    while (temp != nullptr) {
        int val = s.top();
        s.pop();
        if (val != temp->data) {
            return false;
        }
        temp = temp->next;
    }

    return true;
}
